pkg/cmd/cli/cmd/login/helpers.go

```diff
@@ -0,0 +1,139 @@
+package login
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/user/api"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	kclientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
+
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/util/term"
+)
+
+// getMatchingClusters examines the kubeconfig for all clusters that point to the same server
+func getMatchingClusters(clientConfig restclient.Config, kubeconfig clientcmdapi.Config) sets.String {
+	ret := sets.String{}
+
+	for key, cluster := range kubeconfig.Clusters {
+		if (cluster.Server == clientConfig.Host) && (cluster.InsecureSkipTLSVerify == clientConfig.Insecure) && (cluster.CertificateAuthority == clientConfig.CAFile) && (bytes.Compare(cluster.CertificateAuthorityData, clientConfig.CAData) == 0) {
+			ret.Insert(key)
+		}
+	}
+
+	return ret
+}
+
+// findExistingClientCA returns *either* the existing client CA file name as a string,
+// *or* data in a []byte for a given host, and true if it exists in the given config
+func findExistingClientCA(host string, kubeconfig clientcmdapi.Config) (string, []byte, bool) {
+	for _, cluster := range kubeconfig.Clusters {
+		if cluster.Server == host {
+			if len(cluster.CertificateAuthority) > 0 {
+				return cluster.CertificateAuthority, nil, true
+			}
+			if len(cluster.CertificateAuthorityData) > 0 {
+				return "", cluster.CertificateAuthorityData, true
+			}
+		}
+	}
+	return "", nil, false
+}
+
+// dialToServer takes the Server URL from the given clientConfig and does a TCP
+// dial to make sure the server is reachable
+func dialToServer(clientConfig restclient.Config) error {
+	// load CA data in case we have it on a file
+	caData := clientConfig.CAData
+	if len(caData) == 0 && len(clientConfig.CAFile) > 0 {
+		loadedCAData, err := ioutil.ReadFile(clientConfig.CAFile)
+		if err != nil {
+			return err
+		}
+		caData = loadedCAData
+	}
+
+	serverURL, err := url.Parse(clientConfig.Host)
+	if err != nil {
+		return err
+	}
+
+	// try a tcp connection to check if server is reachable
+	timeout := 30 * time.Second
+	switch serverURL.Scheme {
+	case "https":
+		tlsConfig := &tls.Config{}
+		rootCAsN := 0
+		if len(caData) > 0 {
+			pool := x509.NewCertPool()
+			if !pool.AppendCertsFromPEM(caData) {
+				return fmt.Errorf("could not use the provided CA cert file/data")
+			}
+			tlsConfig.RootCAs = pool
+			rootCAsN = len(pool.Subjects())
+		}
+		dialer := &net.Dialer{Timeout: timeout}
+		glog.V(8).Infof("Dialing TCP connection to %s with TLS and %d root CA's", serverURL.Host, rootCAsN)
+		conn, err := tls.DialWithDialer(dialer, "tcp", serverURL.Host, tlsConfig)
```

Comment:
that would also let you do `restclient.TransportFor(&config)` to get the transport, and let it load the CAFile, etc. We probably want to do that anyway, so our debugging hooks wrap the transport and let us see the test dial being done in debug mode